kernel/src/kernel.rs

```diff
@@ -61,6 +62,70 @@ pub struct Kernel {
     grants_finalized: Cell<bool>,
 }
 
+/// Represents a slot for a process in a [ProcessArray].
+///
+/// Holds both a refeference to the [process::Process] as well as a
+/// cached process identifier, which enables efficient selection as
+/// well as invalidating a process without yet deallocating the
+/// process itself.
+///
+/// # Invariants
+///
+/// If [valid_proc_id] is not [ID_INVALID], it should have the same value as [proc_ref]'s [ProcessId#identifier].
+#[derive(Clone)]
+pub struct ProcEntry {
+    /// The [ProcessId#identifier] for the process in [prof_ref], or
+    /// [ID_INVALID] if this [ProcEntry] does not represent an active
+    /// process.
+    pub(crate) valid_proc_id: Cell<usize>,
+    /// Optionally points to a process.
+    ///
+    /// If [valid_proc_id] is not [ID_INVALID], this must be a [Some].
+    pub(crate) proc_ref: Cell<Option<&'static dyn process::Process>>,
+}
+
+impl ProcEntry {
+    pub(crate) const ID_INVALID: usize = !0usize;
+
+    /// Return the underlying [process::Process] if and only if [self]
+    /// represents an active process.
+    pub fn get_active(&self) -> Option<&'static dyn process::Process> {
+        if self.valid_proc_id.get() != Self::ID_INVALID {
+            self.proc_ref.get()
+        } else {
+            None
+        }
+    }
+
+    pub fn get_with_id(&self, identifier: usize) -> Option<&'static dyn process::Process> {
```

Comment:
```suggestion
    pub fn get_if_matches(&self, identifier: usize) -> Option<&'static dyn process::Process> {
```

I also think it would be better to take a `ProcessId` here so that more code isn't inspecting the `ProcessId` type. Or, this should be implemented on the array and not the element.

Reply:
Actually, I don't think this is used and can just be deleted.

Reply:
It is not used